[WIPBetter input set generator and sets for io.lammps

vir-k01 opened the pull request:
## Summary

Major changes:

- feature 1: ...
- fix 1: ...

## Todos

If this is work in progress, what else needs to be done?

- feature 2: ...
- fix 2:

## Checklist

- [ ] Google format doc strings added. Check with `ruff`.
- [ ] Type annotations included. Check with `mypy`.
- [ ] Tests added for new features/fixes.
- [ ] If applicable, new classes/functions/modules have [`duecredit`](https://github.com/duecredit/duecredit) `@due.dcite` decorators to reference relevant papers by DOI ([example](https://github.com/materialsproject/pymatgen/blob/91dbe6ee9ed01d781a9388bf147648e20c6d58e0/pymatgen/core/lattice.py#L1168-L1172))

Tip: Install `pre-commit` hooks to auto-check types and linting before every commit:

```sh
pip install -U pre-commit
pre-commit install
```

DanielYang59 commented on src/pymatgen/core/periodic_table.py:
```diff
@@ -776,6 +777,19 @@ def is_rare_earth(self) -> bool:
         return self.is_lanthanoid or self.is_actinoid or self.symbol in {"Sc", "Y"}
 
+    @property
+    @deprecated(
```
I guess reverting changes to `core.periodic_table` and `io.exciting` is not intended?

esoteric-ephemera replied:
Just behind the master branch, synced up now

gbrunin commented on src/pymatgen/io/lammps/generators.py:
```diff
@@ -10,23 +10,448 @@
 
 MODULE_DIR = os.path.dirname(os.path.abspath(__file__))
 TEMPLATE_DIR = f"{MODULE_DIR}/templates"
+_BASE_LAMMPS_SETTINGS = {
```
To address the different default settings for solids or molecules, maybe this could be a nested dictionary like 
```
_COMMON_SETTINGS = {}  # Add any default that would be common to any type of data. 
_BASE_LAMMPS_SETTINGS = {
    "structure": _COMMON_SETTINGS | {"atom_style": "atomic", "boundary": ("p", "p", "p")},
    "molecule": _COMMON_SETTINGS | {"atom_style": "full", "boundary": ("f", "f", "f")},
    "general": _COMMON_SETTINGS,
}
```
that would then be used according to the type of the object passed to `get_input_set`.

gpetretto replied:
Just as a small comment, I wanted to point out that in principle the `Lattice` has a `pbc` attribute:
https://github.com/materialsproject/pymatgen/blob/39308ec3b5d9bffdaa19ca63a62af2536311e390/src/pymatgen/core/lattice.py#L48
so in principle even mixed use cases could be handled directly with a `Structure` object.
Maybe it would be easier to have the `boundary` and `atom_style` settings be part of the `data` as produced by `LammpsData.from_structure`(`from_molecule`) instead? Would it make sense?

vir-k01 replied:
Thank you so much for the comments! I've made some of the changes, and let me try what @gpetretto mentioned for the molecules

gbrunin commented on src/pymatgen/io/lammps/generators.py:
```diff
@@ -10,23 +10,448 @@
+        input_str = Template(self.inputfile.get_str()).safe_substitute(**settings_dict)  # type: ignore[union-attr]
+        lines = input_str.split("\n")
+        # Filter out the lines where the substitution resulted in a line starting with '#'
+        filtered_input_str = "\n".join([line for line in lines if not line.lstrip().startswith("#")])
```
This may remove some useful comments/stage names that have been written by the user or in the template and that should stay there.
I would suggest to replace missing flags by "###" and fitler the lines starting with "###" (assuming that comments start with a single #).

gbrunin commented on src/pymatgen/io/lammps/generators.py:
```diff
@@ -10,23 +10,448 @@
+        input_file = LammpsInputFile.from_str(filtered_input_str, keep_stages=self.keep_stages)
+
+        return LammpsInputSet(
+            inputfile=input_file, data=data, calc_type=self.calc_type, additional_data=write_data, **kwargs
```
It would be nice to pass the argument `template_file=...` so that a trace of the initial template is kept somehow. This could probably be achieved in different ways though, not sure what would be best.

gbrunin commented on src/pymatgen/io/lammps/generators.py:
```diff
@@ -10,23 +10,448 @@
+
+    def get_input_set(
+        self,
+        data: Structure | LammpsData | CombinedData,
```
This would have to accept Molecule as well.

gbrunin commented on src/pymatgen/io/lammps/generators.py:
```diff
@@ -10,23 +10,448 @@
+                for key, value in base_settings.items():
+                    if key not in settings:
+                        settings[key] = value
+            self.settings = LammpsSettings(validate_params=self.validate_params, **settings)
```
Just a suggestion, but these few lines could be replaced by
```
settings = self.settings.copy()
if self.include_defaults:
    settings = base_settings | settings
self.settings = LammpsSettings(validate_params=self.validate_params, **settings)
```

gbrunin commented on src/pymatgen/io/lammps/generators.py:
```diff
@@ -10,23 +10,448 @@
+        species = ""
+        if isinstance(data, Structure):
+            species = " ".join({s.symbol for s in data.species})
+            data = LammpsData.from_structure(data, atom_style=atom_style)
```
For Molecules, I think the `LammpsData.from_molecule` classmethod should be implemented.

gbrunin reviewed (commented):
Hi Vir,
Thanks for opening this PR, it looks very good and much better than what we had already done. I have left a few comments - those are only suggestions, I don't have a strong opinion on any of them.

I think indeed including support for molecules would be beneficial. I have added a few comments/suggestions going in that direction, let me know what you think.

The force field description is definitely a challenge with so many different ways to specify it. I'm not sure checking/validating what the user specifies is absolutely necessary at this stage - after all the code is also not checking the validity of other user-specified settings. I would leave it to the user to know how to specify the force field they want to use. But if others have ideas on how to implement that efficiently, I'm definitely not against it either :-)

Looking forward to having this integrated in pymatgen!

Best,
Guillaume

vir-k01 commented:
@davidwaroquiers apologies for slacking off on this. After a bit of experimentation, I've added a `LammpsData.from_molecule` method and added different defaults as per what @gbrunin suggested. Tbh although this supports using molecules, a better way would be for the user to first make the `Topology` and `ForceField` objects as they've been currently implemented and use the `LammpsData.from_ff_and_topologies` method instead. This would preserve information about bonding. I suppose this shouldn't be an issue though for force fields that do not rely on such info.

esoteric-ephemera commented:
@mkhorton and @shyuep when you get a chance, would you be able to review this PR? We'd need a new pymatgen release to get @vir-k01's LAMMPS workflows tested on the atomate2 side. Thanks!

@vir-k01's addressed the earlier review comments from community members
